pkg/chunkenc/symbols_test.go

```diff
@@ -175,3 +175,206 @@ func TestSymbolizer(t *testing.T) {
+		},
+		{
+			name: "same string as name and value",
+			labelsToAdd: []labels.Labels{
+				{
+					{Name: "foo-bar", Value: "foo-bar"},
+					{Name: "fizz-buzz", Value: "fizz-buzz"},
+				},
+			},
+			expectedLabels: []labels.Labels{
+				{
+					{Name: "foo_bar", Value: "foo-bar"},
+					{Name: "fizz_buzz", Value: "fizz-buzz"},
+				},
+			},
+			description: "only normalize when string is used as a name, not as a value",
+		},
+		{
+			name: "reused normalized names",
+			labelsToAdd: []labels.Labels{
+				{
+					{Name: "foo-bar", Value: "value1"},
+				},
+				{
+					{Name: "foo-bar", Value: "value2"},
+				},
+			},
+			expectedLabels: []labels.Labels{
+				{
+					{Name: "foo_bar", Value: "value1"},
+				},
+				{
+					{Name: "foo_bar", Value: "value2"},
+				},
+			},
+			description: "normalized names should be cached and reused",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			// Test direct addition
+			s := newSymbolizer()
+			for i, labels := range tc.labelsToAdd {
+				symbols := s.Add(labels)
+				result := s.Lookup(symbols, nil)
+				require.Equal(t, tc.expectedLabels[i], result, "direct addition: %s", tc.description)
+			}
+
+			// Test serialization/deserialization via checkpoint
+			buf := bytes.NewBuffer(nil)
+			_, _, err := s.CheckpointTo(buf)
+			require.NoError(t, err)
+
+			loaded := symbolizerFromCheckpoint(buf.Bytes())
+			for i, labels := range tc.labelsToAdd {
+				symbols := loaded.Add(labels)
+				result := loaded.Lookup(symbols, nil)
+				require.Equal(t, tc.expectedLabels[i], result, "after checkpoint: %s", tc.description)
+			}
+
+			// Test serialization/deserialization via compression
+			buf.Reset()
+			_, _, err = s.SerializeTo(buf, compression.GetWriterPool(compression.Snappy))
+			require.NoError(t, err)
+
+			loaded, err = symbolizerFromEnc(buf.Bytes(), compression.GetReaderPool(compression.Snappy))
+			require.NoError(t, err)
+			for i, labels := range tc.labelsToAdd {
+				symbols := loaded.Add(labels)
+				result := loaded.Lookup(symbols, nil)
+				require.Equal(t, tc.expectedLabels[i], result, "after compression: %s", tc.description)
+			}
+		})
+	}
+}
+
+func TestSymbolizerNormalizationCache(t *testing.T) {
+	s := newSymbolizer()
+
+	// Add a label with a name that needs normalization
+	labels1 := labels.Labels{{Name: "foo-bar", Value: "value1"}}
+	symbols1 := s.Add(labels1)
+
+	// Look up the label multiple times
+	for i := 0; i < 3; i++ {
+		result := s.Lookup(symbols1, nil)
+		require.Equal(t, "foo_bar", result[0].Name, "normalized name should be consistent")
+		require.Equal(t, "value1", result[0].Value, "value should remain unchanged")
+	}
+
+	// Add the same label name with a different value
+	labels2 := labels.Labels{{Name: "foo-bar", Value: "value2"}}
+	symbols2 := s.Add(labels2)
+
+	// The normalized name should be reused
+	result := s.Lookup(symbols2, nil)
+	require.Equal(t, "foo_bar", result[0].Name, "normalized name should be reused")
```

Comment:
This isn't really testing reusing the entry (it's just doing normal string value comparison); you should get the pointer of each with [unsafe.StringData](https://pkg.go.dev/unsafe#StringData) here and compare those.